### ISSUE TITLE
Menu improvement request/idea

### ISSUE THREAD

**Issue by mick-p1982** [FreeGLUTProject/freeglut#52]
Typical 3-D applications of GLUT require rotating and dragging a 3-D model or scene. Usually all buttons are needed for this activity.

On the other hand, clicking the mouse is rarely needed. Therefore if the menu system opened the menu on a click instead of when the button goes down (i.e. with a timer or simple drag logic) then it would be possible to combine context menus with drag based interactions.

(Also it would be very good to have a text input system, for entering things like URLs into a location bar. If that's not asking too much. The menu request seems very sensible doesn't it? If click detection is too much, then how about an API for manually opening a menu under the cursor? That is without attaching it to a button... I don't know why it didn't work like that to begin with myself.)

**Comment by rcmaniac25** [comment]
If I've understood correctly: you want floating menus. This is a reasonable thing to request, though we aren't very quick in development so if you need it soon, you may want to develop it yourself.

Even if we did get to it relatively soon, we'd need to do some number‎ of work to update the relatively old menu system (it's still using direct OpenGL commands rather then GL buffers and shading) and hard codes location and format of menu. I'd think it should be an extension function for creating a floating menu and a function to invoke it's display at a specific location (you notice mouse input, do a display menu call for a menu ID and a position). Along with how the drawing happens, mouse input components would need to be updated. Ensuring this all works on Windows and Linux would be needed. And probably a number of other areas would need to be covered.

So again, you probably want to ‎do this yourself right now. But it's a good suggestion.

**Comment by mick-p1982** [comment]
No. The way classic GLUT is a context menu. The only problem is the menus appear when the mouse is pressed down. Not when the mouse is clicked. A click is down and then up without movement. I think this is easy to understand. Firefox and IE show menus after a full click. 

I'm sorry, I haven't switched to FreeGLUT, but I just assume it still works like classic GLUT. If it already does this, I may switch to it in the short term for a project. I just thought I would share the idea. Because it seems like such a stupid think to hamper menus this way when the library is primarily for spinning 3-D models around :) Thanks for replying!

**Comment by rcmaniac25** [comment]
Oh, so you just want the menu to be invoked on "up/down" instead of just down?

One of the goals of FreeGLUT is to duplicate GLUTs functionality as close as possible. It has additional functions in a separate header that adds onto what GLUT allows. So I'm under the impression that it continues to operate only on mouse down and not down/up.

**Comment by mick-p1982** [comment]
Just as a reminder. GLUT is to make portable applications, for small demos. So "do your own menus" doesn't make any sense at all. I do not believe any application would be harmed if a click opens a menu. And it can be enabled by an option.

Are you a developer? Think of the long term utilty of the package. Dragging is for changing the view of the 3-D demo. Clicking is for opening menus. Dragging has no role to play in menus.

**Comment by dcnieho** [comment]
Vinnie is simply saying that you could implement it in FreeGLUT yourself if you want it soon, we'd of course review and be happy to merge (if the changed behaviour is optional and default off). Its a reasonable suggestion, not soon for us means at least a year from now, quite possibly much longer as we're entirely volunteer driven.

I tried on Chrome here, and indeed a menu come up only when the right mouse button is released, and at the position the cursor is at that point, not where it was originally pressed. You seem to be asking for a menu to not come up at all when the mouse was moved? That would be very user unfriendly (and introduce more state in the freeglut library), clicks usually lead to movement of the cursor of at least a few pixels. 

Feel free to discuss your suggestion on the freeglut developer mailing list, if you plan to implement it, then we can steer you a bit in how it would best fit with the existing code base.

**Comment by mick-p1982** [comment]
That had not occurred to me. I really just meant to provide a common sense tip. I don't expect to keep using GLUT for the project that brought me here or if I want a dependency on a modified version of it.

I was just thinking, if I was preparing a 3-D demo, this would be an improvement. 

One approach that would be easy to implement is to add a callback for suppressing the menu. It could return a "wait" or "rise" or "kill" state. And it could change/configure the active menu if it needed to.

**Comment by rcmaniac25** [comment]
Can you maybe make a list of all the ideas/improvements you have related to menus?

**Comment by mick-p1982** [comment]
The core idea is to be able to manipulate a 3-D model by dragging, but also not have to sacrifice a mouse button by binding it to a menu.

I don't want to over complicate it. I am working on an important project. It's a reference implementation of COLLADA which implements a DOM (Document Object Model) which was one of the original efforts around COLLADA that began at Sony before it was given to Khronos to use as a standard. Things did not go so well for COLLADA by most estimates. I'm kind of having to roll up my sleeve and do the job right.

Implementing XML with C++ is hard it turns out. I've implemented it so that the syntax of C++ itself is used to manipulate the documents so you get a very high level scripting experience and algorithms can be very productive.

There was originally some packages that interpreted the documents and generated visualizations for them. You could say it's a very big demo. Maybe it could get by with GLUT in its most basic form. If you'd like it to showcase FreeGLUT. But at a minimum it requires a location bar. Preferably one that drops down like when the Alt button is pressed for typing a URL into. And also a dropdown menu on the location bar for history. And it could use a context-menu like GLUT provides. But it cannot sacrifice any mouse buttons. And I'm unwilling to render hitboxes for clicking on to access the context menu.

If all of that can be done :) And there's a nice looking FreeGLUT icon, we can use it for the application's icon as far as I know. https://www.khronos.org/collada/wiki/ColladaDOM_3 is the project. Another related project is being developed that falls way outside of GLUT's capabilities, and I thought we'd use the same UI for both, but really it'd be better to limit this simpler application to GLUT if possible.

_It may prove popular because I intend to use it as the first easily accessible demo of a technique I developed almost by accident in 2015 that makes aliasing a thing of the past and produces images that are like television, without jagged edges, and without image-based antialiasing techniques, with no overhead. It sounds miraculous but nobody has really demonstrated any interest in it. There's a deficit of intellectual curiosity in this field for certain. The technique could've been discovered in the 90s and graphics could've been completely different (imagine if photography had jagged edges. Would artists and audiences see it the same way? Or would it not really be taken seriously?)_

**Comment by mick-p1982** [comment]
P.S. I'm worried about Cocoa if we switched to FreeGLUT (it's currently using GLUT) but honestly I've been speaking with an interested party who is offering to help with Cocoa anyway, so I'm confident they will add Cocoa to FreeGLUT if I ask them too. That's something I can offer in return for these features :)

**Comment by rcmaniac25** [comment]
Sorry for the delay...

Let me try again: So you are looking for:

1. Ability to disable menu-invocation so it doesn't eat a click.
2. Menu clicks to be _down-up_ instead of just _down_.
3. A location bar (similar to a browser navigation bar, where the URL is)

If I got that correct, I can tell you the following:

1. I have two thoughts on this: it can either be a callback with a response, or a menu setting where you can disable the menu option. Option three: both. This would be something to discuss on the developer mailing list as we still need to maintain compatibility, so no callback/default enabled state = current behavior of eating click and invoking menu.
2. I can see this being a global setting where the default is to invoke on the down-click. Again, something to discuss on the developer mailing list.
3. You will have to implement this yourself. While there is a menu system, we don't have a UI framework. The menu system also needs some work, as stated previously, so that it's still GLUT compatible but also works on mobile and probably updated in general to look better (or something similar).

Cocoa support would be greatly appreciated if the party you mentioned is willing to do so. I would recommend them posting to our developer mailing list so we can discuss the work instead of it being completed and us taking issue with different aspects of implementation.

As you can tell, discussing this on the developer mailing list is highly recommended. In addition, if you want this functionality sooner rather then later, please contribute as it will produce faster results for you then waiting for us to get to the requests and it helps improve FreeGLUT.

**Comment by mick-p1982** [comment]
My reason for suggesting the menu pop-up control callback is to make "2. Menu clicks to be down-up instead of just down." a non-issue. Since that was a concern.

The other reason is that it's something that I could conceivably contribute myself. Since it's a portable solution. Or should be. (Assuming it's no problem with round trips in remote X.)

But I can't contribute a location-bar. That will touch on more systems than I am comfortable with. It would require someone who is familiar with all of portability targets to oversee it. I can't implement it on all targets. And it has to be tested at a minimum. I'm not interested in partial support.

Implementing Cocoa should be relatively simple because it's isolated to one target (Cocoa.)

I can only offer a trade. And excuse me. But I really cannot stand mailing lists. So that might be the end of this. I'm glad to hear back from you though. And appreciate the correspondence. I'd like to see FreeGLUT make some very practical changes.

I'm available via email if we can arrange something. Or I will use a discussion forum if there is one. (This project is going well. I think it needs to support widgets for Cg like annotations. But I think maybe they can be modeled in 3-D and appear in accessory windows, as an excuse for them to not use the native UI and to use the same familiar inputs as the model-viewer window.)

**Comment by dcnieho** [comment]
Sorry for the very slow reply. Academic here with many deadlines.
Juni-August will be much more free for me and I'll respond with a
normal delay then.

As Vinnie said, FreeGLUT indeed does not have much in the way of UI,
if anything. It has menus, which is hardly more than a bit of hackery
and old API opengl calls. I fully agree with the need for any
additional things to be available for all platforms, this is paramount
for FreeGLUT. That said, the location bar may be best implemented in
user code, and not in FreeGLUT itself.

Only having menu's appear on down-up instead of immediately on down is
something we could do by means of a global (default off) setting. It
is a nice addition, menus that appear on down are a bit jarring, but
backward compatibility will keep that as a the default for us. This
should be trivial to implement. If you would want to have a crack at
it, we'd love a pull request and any discussion you would like. see
fgCheckActiveMenu() in fg_menu.c.

Implementing this though a menu popup callback may be less trivial.
That callback would replace part of the logic in fgCheckActiveMenu()
(probably the lowest part checking if a menu should be opened), and a
default callback would implement exactly that logic. What should the
parameters to the callback be? simply the same as fgCheckActiveMenu()
itself takes, suitable converted to user-facing APIs (e.g. window
number)? I'd be happy to discuss this further.

Cocoa support is something we're very interested in and be happy to facilitate!

What would you like to trade?

P.S. what is this "gone with the jaggy edges" technique? Any links to
info about it? Sounds interesting indeed, especially if its that
cheap.

**Comment by mick-p1982** [comment]
Hi dcnieho,

As it happens today I put up a first sample of this new COLLADA reference-implementation prebuilt for Windows only:

 https://sourceforge.net/p/collada-dom/discussion/531263/thread/05af0139/

Right now I'm thinking about either working with FreeGLUT some or restoring "ZAE" to this beta version of the library which is the compressed/archive specification for COLLADA. 

I will propose what I think works best for menus and possibly implement it unless I run up against non-portable code regions.

A location bar is my sticking point. It cannot be implemented in nonportable code, and it really should use a native text input box on all supported platforms. It's ambitious, but I think FreeGLUT should strongly consider pursuing text/URI input because we are in the network age. Text is really very complex, which is why it needs to be native, and it cannot be optional. I think a single line input box with an autocomplete style callback is ideal, just for displaying a history, populated like a menu.

I would "like to trade" work on minimal feature support, possibly for a Cocoa coder who has expressed interest. Try to find the right mix between cross-platform and barely functional.

P.S. Here ( http://swordofmoonlight.net/holy/Moratheia(6-PNGs).zip ) is a link with sample images without aliasing and full-screen straight diagonals. I'm surprised I did not share the link before. It might be newer than the comment, but I don't think so. There are several complementary effects. The most interesting one modulates subpixel rounding so that the knots of the jagged lines fall on a different pixel in each frame and so disappear in superposition. The effect is lost somewhat in a still image, but there's a two frame dissolve also for best picture and it gets close to the full-motion experience. The subpixel stuff is done in the vertex shader for free, producing better than MSAA style results (with no buffers) and the dither and other kinds of effects do not do image-processing, meaning there is a 1:1 relationship where sampling textures (no sampling around the edges of texels.) The subpixel technique would be better managed/explored in hardware I think. It's unknown or not wellknown to the degree that software that should use it does not use it (virtually all real-time 3-D software should provide the option if not use it as a default/showcase mode.) There's nothing in it that is high-tech. It could've been around since the beginning of 3-D video games and things I think. I may be repeating the other comment by now.

**Comment by mick-p1982** [comment]
I think I'm about to look at FreeGLUT. I took some time to work on COLLADA's spline (curves) and related animation methods because it seemed like a loose end to me. 

Here (below) is I believe two sketches for how to do a text input box with xlib which is my main concern. I really don't want to be responsible for implementing a text box as a project outsider, but I was curious if it was even practical or not. It seems completely doable to me:

https://github.com/vlnx/xim-example/blob/master/xim.c
https://gist.github.com/baines/5a49f1334281b2685af5dcae81a6fa8a

I need someone closer to FreeGLUT to take an interest in it. I can only continue to use GLUT if it has a single line text box. A COLLADA reference implementation is about as a prestigious as a GLUT application gets, so please work with me here. The time table is not as important as the resolution to make it cross platform. I will work on the Win32 version of the feature, so there is a model. I don't want to shoulder more than that.

**Comment by jtsiomb** [comment]
This guy is clearly retarded, absorbed in his self-importance, and unwilling to contribute other than give directions. "A COLLADA reference implementation is prestigious" ?!  What the actual fuck does that even mean? A technique he developed that "makes aliasing a thing of the past"?! "MAKE A TRADE"? That's not how free software projects work my friend...

I'm just passing by, so feel free to continue discussing this if you like, but I really don't see why you keep talking to this guy.

At the very least, all messages not directly about the feature request should be deleted (including this message really). Discussing collada implementations and antialiasing techniques have no place in a freeglut issue tracker, and just add useless noise.

**Comment by mick-p1982** [comment]
I think we are just replying to each other jtsiomb. We apologize if you are receiving email notifications you do not desire. I just found out yesterday that the SF.net is the real home of this project, so I am moving my work to the discussion boards there, if that's any consolation to you.

**Comment by mick-p1982** [comment]
CORRECTION: There aren't discussion boards at SF.net, but I did prepare a PATCH https://sourceforge.net/p/freeglut/feature-requests/88/ in relation to this feature request...

WELL to be honest, I never requested this feature. It was just a thought I had. That I shared because I thought GLUT users would benefit. I did not expect to strike up a conversation. Because rcmaniac25 and dcnieho were available they caused me to reconsider.

**Comment by jtsiomb** [comment]
This is not a place for conversations. This is a bug/issue tracking database.

**Comment by mick-p1982** [comment]
I came here via a link on the Free GLUT website requesting feature requests. If a feature is requested it must be understood by the maintainers, and I think that is what they are interested in discussing. Of course, it's not my place, as an outsider, to not reply to maintainers or anyone :)

(The link to this webpage has the word "feature-request" in the URL.)

**Comment by nigels-com** [comment]
An alternative that comes to find for text editing and menus: https://github.com/libglui/glui
